beacon-chain/sync/subscription_topic_handler.go

```diff
@@ -0,0 +1,90 @@
+package sync
+
+import (
+	"sync"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p"
+)
+
+// This is a subscription topic handler that is used to handle basic
+// CRUD operations on the topic map. All operations are thread safe
+// so they can be called from multiple routines.
+type subTopicHandler struct {
+	sync.RWMutex
+	subTopics map[string]*pubsub.Subscription
+	digestMap map[[4]byte]int
+}
+
+func newSubTopicHandler() *subTopicHandler {
+	return &subTopicHandler{
+		subTopics: map[string]*pubsub.Subscription{},
+		digestMap: map[[4]byte]int{},
+	}
+}
+
+func (s *subTopicHandler) addTopic(topic string, sub *pubsub.Subscription) {
+	s.Lock()
+	defer s.Unlock()
+	s.subTopics[topic] = sub
+	digest, err := p2p.ExtractGossipDigest(topic)
+	if err != nil {
+		log.WithError(err).Error("Could not retrieve digest")
+		return
+	}
+	s.digestMap[digest] += 1
+}
+
+func (s *subTopicHandler) topicExists(topic string) bool {
+	s.RLock()
+	_, ok := s.subTopics[topic]
+	s.RUnlock()
+	return ok
+}
+
+func (s *subTopicHandler) removeTopic(topic string) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.subTopics, topic)
```

Comment:
better to  delete this after the digest is decremented succesfully.

Reply:
Slightly trickier than that as the 2 maps are technically unrelated, 
`subTopics` tracks active subscriptions and
`digestMap` tracks the number of topics registered for a particular digest. 

Even if `digestMap` doesn't have the appropriate digest mapped(should not be possible), we should still remove it from the subscription map. Which is why i remove it first.